Feature : Enable LLDB Debugging for R Dev Container

avinab-neogy opened the pull request:
Added lldb debugging functionality

Closes https://github.com/r-devel/r-dev-env/issues/196 and https://github.com/r-devel/r-dev-env/issues/30

avinab-neogy committed:
changes to devconatiner.json

avinab-neogy committed:
added launch configuration and changed script to move the file to .vscode inside the container

avinab-neogy committed:
made a correction to the referencing of the launch.json file in the script

avinab-neogy committed:
added step to set cflags in building r tutorial

avinab-neogy committed:
added allow_ptrace script

avinab-neogy committed:
added launch_r script to load the wrapper

avinab-neogy committed:
updated local script to build the library and make the wrapper executable

avinab-neogy committed:
updated which_r to execute launch_r and update it

avinab-neogy committed:
removed vscode,devocntainer and scripts dir from files exclude beacuse the scripts need to be accessed from the container

pre-commit-ci[bot] committed:
[pre-commit.ci] Fixing issues with pre-commit

hturner commented on .devcontainer/devcontainer.json:
```diff
@@ -9,8 +10,6 @@
             "settings": {
                 "git.ignoredRepositories": ["."],
                   "files.exclude": {
-                       "**/.vscode": true,
-                       "**/.devcontainer": true,
```
These folders were deliberately hidden from the user in #265 , you can still access them directly, e.g. open `settings.json` via `code .vscode/setting.json`.

However, as the user may wish to edit the VS Code settings, it is fair enough to keep that visible, so let's keep L12 deleted but reinstate L13.

hturner commented on .devcontainer/devcontainer.json:
```diff
@@ -19,16 +18,15 @@
                        "**/.pre-commit-config.yaml": true,
                        "**/Dockerfile": true,
-                       "**/mkdocs.yml": true,
-                       "**/scripts": true
```
Similarly, please reinstate `"**/scripts": true`

hturner commented on .devcontainer/devcontainer.json:
```diff
@@ -19,16 +18,15 @@
                 "r.plot.useHttpgd": true,
                 "r.rpath.linux": "/usr/bin/R",
-                "r.rterm.linux": "/usr/bin/R",
+                "r.rterm.linux": "/workspaces/r-dev-env/scripts/launch_r.sh",
```
The root directory may not be `/workspaces/r-dev-env/`, so let's have this as 
```suggestion
                "r.rterm.linux": "/usr/bin/R",
```
to start with and use `localscript.sh` to update this to `"$WORK_DIR/scripts/launch_r.sh"` (it must be a full path, so we can't simply use `./scripts/launch_r.sh` here)

hturner commented on scripts/launch_r.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+export LD_PRELOAD=/workspaces/r-dev-env/scripts/allow_ptrace.so
+# exec /workspaces/r-dev-env/build/r-devel/bin/R "$@"
```
The root directory may not be `/workspaces/r-dev-env/`. A relative path works for the `export` command and the other is just a comment so we can update this to

```suggestion
export LD_PRELOAD=./scripts/allow_ptrace.so
# `which_r.sh` sets the R binary used in the command below
```

hturner commented on scripts/which_r.sh:
```diff
@@ -3,6 +3,7 @@
 which_r() {
   # Specify the parent directory
   parent_dir="$WORK_DIR/build"
+  launch_script="$WORK_DIR/scripts/launch_r.sh"
```
This line does not match the comment above. You can define `launch_script` later, when it is needed, e.g. after L63.

hturner commented on scripts/allow_ptrace.c:
```diff
@@ -0,0 +1,7 @@
+
+__attribute__((constructor)) void allow_ptrace() {
+    prctl(PR_SET_PTRACER, PR_SET_PTRACER_ANY);
+    fprintf(stderr, "allow_ptrace: called\n");
```
Delete this line, this was just used for debugging. (Otherwise we end up with lots of "allow_ptrace: called" print outs when using R!)

hturner reviewed (commented):
This is getting close, but we need to avoid hard-coding the root directory. 

Another place this needs to change is in the postCreateCommand: use `./scripts/localscript.sh` not `/workspaces/r-dev-env/scripts/localscript.sh`.

Also, delete `rm INIT` in `welcome_message.sh` as the INIT file no longer exists (this should have been part of #263 but we can tidy that up here).

hturner commented on docs/tutorials/building_r.md:
```diff
@@ -55,7 +55,22 @@ mkdir -p $BUILDDIR
+(`-O0`) so that the structure of the code closely matches the
+original source.
+
+**6) Configure the build**
```
Testing this more carefully, I found that the CFLAGS setting needs to be on the same line as configure. So we need to delete this optional step and get everyone to configure R without compiler optimisations for C - this is fine for the dev container where it is designed for working on the source code.

The new configure command will be

```bash
$TOP_SRCDIR/configure --with-valgrind-instrumentation=1 CFLAGS="-g -O0"
```

avinab-neogy committed:
merged with the latest devel

avinab-neogy committed:
removed absolute paths

avinab-neogy committed:
updated allow ptrace to remove printing multiple times when the wrapper is loaded and updated launch r to use relative path

avinab-neogy committed:
updated building r documentation to include the cflags step into the configure command

avinab-neogy committed:
Update r.rterm.linux setting to use the launch_r.sh script with full dynamic path

avinab-neogy committed:
move the launch_script variable definition in which_r script

avinab-neogy commented:
Hi @hturner ,
I have made all the changes suggested in the review and am testing this out.
Thanks!

avinab-neogy commented:
<img width="2492" height="1444" alt="Screenshot from 2025-08-28 14-36-18" src="https://github.com/user-attachments/assets/9a4550e9-6855-495e-b24d-d6938a00db16" />
Tested out after making the recent changes and it still returns the same error.
<img width="1508" height="877" alt="Screenshot from 2025-08-28 14-36-42" src="https://github.com/user-attachments/assets/3555637f-f00f-4e25-9aa9-d49f6d307166" />

avinab-neogy commented:
Hi @hturner ,
I tested this setup using codespaces and after building R from source following the steps from https://contributor.r-project.org/r-dev-env/tutorials/building_r/ and running `which_r` to select `r_devel` the process is still selecting the default binary at `/usr/bin/R` 
<img width="2330" height="1423" alt="Screenshot from 2025-08-28 18-40-57" src="https://github.com/user-attachments/assets/0b938e42-670c-4f51-a299-64ea686ddf68" />
I am still looking into the behavior of which_r to understand why the selection is not updating the terminal to use the built R binary correctly. Will keep you posted once I have more insights!

hturner commented:
`which_r` changes the R version used in **R terminals**, which is set by the VS Code setting `r.rterm.linux`. With this PR, we are fixing `r.term.linux` to always point to `launch_r.sh` and using `which_r` to update the R binary specified in `launch_r.sh`.

The version of R used in the bash terminal is defined by the path returned by `which r`, where `which` is the usual Linux command. This would be replaced by the r-devel version if we used `make install` after building R. However, we don't want to do this 1) because it would require using sudo, which we shouldn't encourage users to do and 2) because we want to use the release version of R for the packages that we install when setting up  the container (see https://github.com/r-devel/r-dev-env/issues/25#issuecomment-1688103594).

So to check things are working properly:

1) After running `which_r` to select the r-devel version, check that `launch.sh` has been updated as expected (`exec /usr/bin/R "$@"` replaced with `exec /workspaces/r-dev-env/build/r-devel/bin/R "$@"`)
2) Launch an R terminal by clicking "R (not attached)" in the status bar or selecting "R: Create R terminal" through the command palette. The r-devel version of R should be launched (which you should see by the version information at the top of the opening message in R). Run `Sys.getenv("LD_PRELOAD")` and check it is set to `"./scripts/allow_ptrace.so"`. Then LLDB debugging should work!

pre-commit-ci[bot] committed:
[pre-commit.ci] Fixing issues with pre-commit

avinab-neogy commented:
I have updated the relevant section in `which_r.sh` to modify `launch_r.sh` so that it correctly sets the R executable for the R terminal. Using `[^ ]` for pattern matching makes the substitution more precise and reduces the risk of erroneous replacements.

I have updated the `devcontainer.json` to use `bash` instead of `sh` since that was returning an error as well.

avinab-neogy committed:
updated which_r.sh

avinab-neogy committed:
updated devcontainer.json

avinab-neogy committed:
updated devcontainer.json

avinab-neogy commented:
Hi @hturner ,
I know you are busy right now but please check when you are free, I tried building R in codespaces using this `feature/lldb-new` branch and I am geeting this error and this is suggesting that `local_script.sh` script is not executable but I don't get this error in the local setup, what should be done here?
<img width="2117" height="694" alt="image" src="https://github.com/user-attachments/assets/b4d7e5ee-4c96-42ff-8df1-bef2617e1693" />

hturner committed:
make allow_ptrace an executable program

this avoids using LD_PRELOAD environment variable

hturner committed:
ensure VSCode R extension initializes

hturner committed:
tidy .gitignore

pre-commit-ci[bot] committed:
[pre-commit.ci] Fixing issues with pre-commit

hturner committed:
hide .devcontainer and scripts

hturner committed:
delete INIT as no longer used

hturner committed:
update comment

hturner committed:
Merge remote-tracking branch 'origin/devel' into feature/lldb-new

hturner commented:
@avinab-neogy sorry for the long delay in testing this. 

I converted the `allow_ptrace` code to a standalone C program that allows ptrace then executes a given R binary - this avoids using the LD_PRELOAD environment variable to load a shared object file, which caused errors when installing R packages as file could not be found  (the errors were ignorable but still a bit messy).

By setting `r.rterm.linux` to `/workspaces/r-dev-env/scripts/launch_r.sh` it seems that the VS Code R extensions isn't initialized properly, so I added some code to `.Rprofile` to do that explicitly.

Other than that I did some minor tidying and brought the branch in line with `devel`.

This works okay in a local setup and I think should work in codespaces too - the error you mention in https://github.com/r-devel/r-dev-env/pull/267#issuecomment-3268839912 should not be an issue as you updated `which_r.sh` to define the full path to `launch.sh` based on the workspace root. So I'll merge this in and we'll see how it goes.
